blockchain/blockdao/blockdao.go

```diff
@@ -398,7 +398,19 @@ func (dao *blockDAO) GetReceipts(blkHeight uint64) ([]*action.Receipt, error) {
 }
 
 func (dao *blockDAO) PutBlock(ctx context.Context, blk *block.Block) error {
-	err := func() error {
+	// early exit if block already exists
+	blkHeight := blk.Height()
+	blkHash, err := dao.GetBlockHash(blkHeight)
```

Comment:
our coding practice is that, when a struct has lock/mutex, exported method should only call internal un-exported method, but not other exported method (to avoid potential deadlock)

so use getBlockHash() here